frontend/src/metabase-lib/v1/Question.ts

```diff
@@ -433,36 +436,25 @@ class Question {
    * of Question interface instead of Query interface makes it more convenient to also change the current visualization
    */
 
-  composeThisQuery(): Question | null | undefined {
-    if (this.id()) {
-      const card = {
-        display: "table",
-        dataset_query: {
-          type: "query",
-          database: this.databaseId(),
-          query: {
-            "source-table": getQuestionVirtualTableId(this.id()),
-          },
-        },
-      };
-      return this.setCard(card);
+  composeQuestion(): Question {
+    if (!this.isSaved()) {
+      return this;
     }
-  }
 
-  composeDataset(): Question {
-    const type = this.type();
+    const metadata = this.metadataProvider();
+    const tableId = getQuestionVirtualTableId(this.id());
+    const table = Lib.tableOrCardMetadata(metadata, tableId);
+    const query = Lib.queryFromTableOrCardMetadata(metadata, table);
+    return this.setQuery(query);
+  }
 
-    if (type === "question" || !this.isSaved()) {
+  composeQuestionAdhoc(): Question {
```

Comment:
I fail to recognize the difference between `composeQuestionAdhoc` and `composeQuestion`.
Can't we just remove `composeQuestionAdhoc` and use `composeQuestion().composeQuestion()`?

Can you briefly describe it please?

Reply:
`composeQuestionAdhoc` creates an ad-hoc question based on this question. `composeQuestion` replaces the query in the current question "in-place" without making it an ad-hoc question.